src/coreclr/jit/optimizebools.cpp

```diff
@@ -1698,18 +1698,19 @@ bool Compiler::fgFoldCondToReturnBlock(BasicBlock* block)
     }
 
     // Both edges must be BBJ_RETURN
-    BasicBlock* retFalseBb = block->GetFalseTarget();
-    BasicBlock* retTrueBb  = block->GetTrueTarget();
+    BasicBlock* const retFalseBb = block->GetFalseTarget();
+    BasicBlock* const retTrueBb  = block->GetTrueTarget();
 
-    // Although, we might want to fold fallthrough BBJ_ALWAYS blocks first
-    if (fgCanCompactBlock(retTrueBb))
+    // We might want to compact BBJ_ALWAYS blocks first,
+    // but don't compact the conditional block away in the process
+    if (fgCanCompactBlock(retTrueBb) && !retTrueBb->TargetIs(block))
     {
         fgCompactBlock(retTrueBb);
         modified = true;
     }
     // By the time we get to the retFalseBb, it might be removed by fgCompactBlock()
     // so we need to check if it is still valid.
-    if (!retFalseBb->HasFlag(BBF_REMOVED) && fgCanCompactBlock(retFalseBb))
+    if (!retFalseBb->HasFlag(BBF_REMOVED) && fgCanCompactBlock(retFalseBb) && !retFalseBb->TargetIs(block))
```

Comment:
presumably you also don't need `!retFalseBb->HasFlag(BBF_REMOVED)` check

Reply:
I think we still need this too in the case of the diamond shape: If the true target jumps to the false target and we compact them, `retFalseBb` will be removed from the flowgraph. This check currently defends against this case.